sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/complexTypeCreator.scala

```diff
@@ -134,49 +155,81 @@ case class CreateMap(children: Seq[Expression]) extends Expression {
   }
 
   override def doGenCode(ctx: CodegenContext, ev: ExprCode): ExprCode = {
-    val arrayClass = classOf[GenericArrayData].getName
     val mapClass = classOf[ArrayBasedMapData].getName
     val keyArray = ctx.freshName("keyArray")
     val valueArray = ctx.freshName("valueArray")
-    ctx.addMutableState("Object[]", keyArray, s"this.$keyArray = null;")
-    ctx.addMutableState("Object[]", valueArray, s"this.$valueArray = null;")
 
-    val keyData = s"new $arrayClass($keyArray)"
-    val valueData = s"new $arrayClass($valueArray)"
+    val MapType(keyDt, valueDt, _) = dataType
+    val evalKeys = keys.map(e => e.genCode(ctx))
+    val isPrimitiveArrayKey = ctx.isPrimitiveType(keyDt) && evalKeys.forall(_.isNull == "false")
```

Comment:
I theoretically agree with you. The key of a map should not be null.
However, in the generate code, we see the following stuff. Thus, I thought that there are some cases that `eval.isNull` may have null.
```
if (${eval.isNull}) {
  throw new RuntimeException("Cannot use null as map key!");
} else {
  ...
```
If we can understand`eval.isNull` is never `null`, I really like your optimization. Do you have any thought?